src/msg/async/crypto_onwire.cc

```diff
@@ -0,0 +1,295 @@
+// https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption
+// https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+class AES128GCM_OnWireTxHandler : public ceph::crypto::onwire::TxHandler {
+  static constexpr const std::size_t AESGCM_KEY_LEN{16};
+  static constexpr const std::size_t AESGCM_IV_LEN{12};
+  static constexpr const std::size_t AESGCM_TAG_LEN{16};
+  static constexpr const std::size_t AESGCM_BLOCK_LEN{16};
+
+  using nonce_t = std::array<unsigned char, AESGCM_IV_LEN>;
+
+  CephContext* const cct;
+  std::unique_ptr<EVP_CIPHER_CTX, decltype(&::EVP_CIPHER_CTX_free)> ectx;
+  ceph::bufferlist buffer;
+  // using GCC's "Tetra Integer" mode here
+  ceph::uint128_t nonce;
+
+public:
+  AES128GCM_OnWireTxHandler(CephContext* const cct,
+			    const AuthConnectionMeta& auth_meta,
+			    const ceph::uint128_t& nonce)
+    : cct(cct),
+      ectx(EVP_CIPHER_CTX_new(), EVP_CIPHER_CTX_free),
+      nonce(nonce) {
+    ceph_assert_always(auth_meta.connection_secret.length() >= AESGCM_KEY_LEN);
+    ceph_assert_always(ectx);
+
+    if (1 != EVP_EncryptInit_ex(ectx.get(), EVP_aes_128_gcm(),
+			        nullptr, nullptr, nullptr)) {
+      throw std::runtime_error("EVP_EncryptInit_ex failed");
+    }
+
+    if(1 != EVP_EncryptInit_ex(ectx.get(), nullptr, nullptr,
+	reinterpret_cast<const unsigned char*>(auth_meta.connection_secret.c_str()),
+	nullptr)) {
+	//reinterpret_cast<const unsigned char*>(&nonce))) {
+      throw std::runtime_error("EVP_EncryptInit_ex failed");
+    }
+  }
+
+  ~AES128GCM_OnWireTxHandler() override {
+    memset(&nonce, 0, sizeof(nonce));
+  }
+
+  std::uint32_t calculate_segment_size(std::uint32_t size) override
+  {
+    return size;
+  }
+
+  void reset_tx_handler(
+    std::initializer_list<std::uint32_t> update_size_sequence) override;
+
+  ceph::bufferlist::contiguous_filler reserve(const std::uint32_t size) override {
+    return buffer.append_hole(size);
+  }
+
+  void authenticated_encrypt_update(ceph::bufferlist&& plaintext) override;
+  ceph::bufferlist authenticated_encrypt_final() override;
+};
+
+void AES128GCM_OnWireTxHandler::reset_tx_handler(
+  std::initializer_list<std::uint32_t> update_size_sequence)
+{
+  if(1 != EVP_EncryptInit_ex(ectx.get(), nullptr, nullptr, nullptr,
+      reinterpret_cast<const unsigned char*>(&nonce))) {
+    throw std::runtime_error("EVP_EncryptInit_ex failed");
+  }
+
+  buffer.reserve(std::accumulate(std::begin(update_size_sequence),
+    std::end(update_size_sequence), AESGCM_TAG_LEN));
+  // transplantate the nonce update management from OpenVPN's AES-GCM
+  nonce += 1ULL << 32;
+}
+
+void AES128GCM_OnWireTxHandler::authenticated_encrypt_update(
+  ceph::bufferlist&& plaintext)
+{
+  int update_len = 0;
+  auto filler = buffer.append_hole(plaintext.length());
+  if(1 != EVP_EncryptUpdate(ectx.get(),
+	reinterpret_cast<unsigned char*>(filler.c_str()),
+	&update_len,
+	reinterpret_cast<const unsigned char*>(plaintext.c_str()),
```

Comment:
we may eventually want to avoid c_str() on plaintext to allow a gatherlist?

Reply:
Definitely TODO.

Reply:
Fixed.